kvbc/test/sparse_merkle_storage/db_adapter_unit_test.cpp

```diff
@@ -248,23 +357,126 @@ TEST_P(db_adapter_custom_blockchain, add_block_throws_if_deletes_contain_non_pro
   ASSERT_THROW(adapter.addBlock(updates, deletes), std::runtime_error);
 }
 
+TEST_P(db_adapter_custom_blockchain, getValue_finds_overlapping_keys) {
+  const Sliver skey1{"key1"};
+  const Sliver skey2{"key2"};
+  concord::kvbc::v2MerkleTree::DBAdapter::NonProvableKeySet non_provable_key_set = {skey1, skey2};
+
+  auto adapter = DBAdapter{GetParam()->newEmptyDb(), true, non_provable_key_set};
+  const auto numBlocks = 10u;
+  for (size_t i = 0u; i < numBlocks; ++i) {
+    adapter.addBlock(SetOfKeyValuePairs{{skey1, Sliver{"value1" + std::to_string(i)}},
+                                        {skey2, Sliver{"value2" + std::to_string(i)}}});
+  }
+  for (size_t i = 0u; i < numBlocks; ++i) {
+    {
+      const auto &[value, blockId] = adapter.getValue(skey1, i + 1);
+      ASSERT_EQ(value, Sliver{"value1" + std::to_string(i)});
+      ASSERT_EQ(blockId, i + 1);
+    }
+    {
+      const auto &[value, blockId] = adapter.getValue(skey2, i + 1);
+      ASSERT_EQ(value, Sliver{"value2" + std::to_string(i)});
+      ASSERT_EQ(blockId, i + 1);
+    }
+  }
+
+  // Make sure that when the last reachable block id is provided, the most updated value is returned
+  {
+    const auto &[value, blockId] = adapter.getValue(skey1, adapter.getLastReachableBlockId());
+    ASSERT_EQ(blockId, numBlocks);
+    ASSERT_EQ(value, Sliver{"value1" + std::to_string(numBlocks - 1)});
+  }
+  {
+    const auto &[value, blockId] = adapter.getValue(skey2, adapter.getLastReachableBlockId());
+    ASSERT_EQ(blockId, numBlocks);
+    ASSERT_EQ(value, Sliver{"value2" + std::to_string(numBlocks - 1)});
+  }
+}
+
+TEST_P(db_adapter_custom_blockchain, getValue_finds_non_provable_keys) {
```

Comment:
I guess this test checks single-versioned non-provable keys, in contrast to the test above. Maybe we can name it such that it is explicit. Do I miss something?